### ISSUE TITLE
Make "kubectl explain" work for HTTPProxy resources.

### ISSUE THREAD

**Issue by jpeach** [projectcontour/contour#1723]
**Please describe the problem you have**

`kubectl explain` is a built-in way to provide Kubernetes resource reference material to users. It would be great to support this as part of the `HTTPProxy` documentation.

**Comment by jpeach** [comment]
@youngnick I think that `kubectl explain` pull this from the OpenAPI definition for the Resource, so maybe we could get this for free as part of #1575?

**Comment by jpeach** [comment]
```$ k explain httpproxy
KIND:     HTTPProxy
VERSION:  projectcontour.io/v1

DESCRIPTION:
     <empty>
```
```$ k explain httpproxy.spec
KIND:     HTTPProxy
VERSION:  projectcontour.io/v1

DESCRIPTION:
     <empty>
```

:(

**Comment by youngnick** [comment]
Hmm, probably. I'll take this for now and see if that happens when I do #1414.

**Comment by davecheney** [comment]
There is also #1567

**Comment by youngnick** [comment]
Sadly, #1731 didn't just magically fix this one, but I think that that's due to upstream things.

Specifically:
- [Pruning](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#preserving-unknown-fields) must be enabled. This is optional for CRD version `v1beta1` (what we're currently using). We can specify it, but not with the generator I added in #1731, it doesn't support it yet. I enabled it manually in a test 1.16 cluster, but the `kubectl explain` was not the full version,  so I suspect the second condition is also not fulfilled currently.
- The schema must be [structural](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#specifying-a-structural-schema), which the ones we are publishing *should* be (thanks to the generator). I suspect that we may need to update to a newer version of the generator, but master is broken in my version. In addition, I suspect something we are doing in our current version is making the validation not structural, but I can't find what it is.

The tl;dr here for next steps is:
- We need to require at least Kubernetes 1.16, and move to CustomResourceDefinition `v1`. That will make solving this a lot easier. It's a big change though, and will require a long tail for updates.
- Until that happens the best we can do is chase `controller-tools` to add support for setting the pruning value in the spec, and to add support for ensuring the validation is structural.

There is no way any of this will be done in rc.2, so I'm moving this to backlog. I'll keep it and poke it along.

**Comment by jpeach** [comment]
Blocked by https://github.com/projectcontour/contour/issues/2678

**Comment by alexbrand** [comment]
+1
